### PR TITLE
#10 generation skipping some characters

### PR DESCRIPTION
**Pull request by kengoodridge**
This is a workaround for an issue with enigo with strings that start with newline (maybe macos only?)

I made sure it never passes a string to enigo.text("\nbad") that started with a newline

### CODE REVIEW

**Comment by jasonjmcghee** [comment]
I will review this soon- super excited

**Comment by kengoodridge** [comment]
@jasonjmcghee This project is so pretty in its usefulness to size ratio. It's like the perfect solo open source project. I'm so jealous I didn't think of it first. 🤣

**Comment by jasonjmcghee** [comment]
You're going to kill me, but I don't agree with the "Enter" change. That will break a lot of use cases, like typing into an input that triggers submit on enter

**Comment by jasonjmcghee** [comment]
> @jasonjmcghee This project is so pretty in its usefulness to size ratio. It's like the perfect solo open source project. I'm so jealous I didn't think of it first. 🤣

Fwiw, definitely not the first to think of this, people reached out to me that said they built this before / 10 years ago etc.

But I built a prototype and it was immediately useful to me, so I wanted to share! And hopefully find developers like you so we could make the project better together.

**Comment by kengoodridge** [comment]
no killing here. Wasn't super confident with it. Tell me if you can reproduce it tho. 

Feel free to edit that out of this PR. The top part fixes the missing text tho, or seem to anyway.

File: src-tauri/src/main.rs [hunk 1]
```diff
@@ -329,19 +329,21 @@ fn trigger_action(
                         whole_buffer.push(response.clone());
                         delta_buffer.push(response);
 
-                        let delta_output = {
-                            if delta_buffer.len() > 4 {
```
Comment (jasonjmcghee) [inline src-tauri/src/main.rs:333]: I regret not writing comments so much... I have no idea why I did this, and had `rem` disabled b/c I was hacking on it 🤦 . I will try to go spelunk in history...

Comment (kengoodridge) [reply src-tauri/src/main.rs:333]: I can undo the enter thing... got it locally. Want me to push? Definitely skipping new lines for me tho...

list the meanings of life

results in

1. Meaning in a philosophical sense: The purpose or significance of human existence, often explored through existentialism and other philosophies.2. Scientific meaning: Biological functions such as reproduction and survival.3. Personal meaning: Subjective understanding and personal beliefs about one's role and place in the world.4. Spiritual meaning: Interpretations derived from religious or spiritual beliefs, often related to connection with a higher power, enlightenment, or spiritual growth.5. Collective meaning: Societal values, norms, and cultural traditions that shape our understanding of life's purpose.

**Comment by jasonjmcghee** [comment]
> no killing here. Wasn't super confident with it. Tell me if you can reproduce it tho.
> 
> Feel free to edit that out of this PR. The top part fixes the missing text tho, or seem to anyway.

I'm trying to understand why your changeset fixes it. The buffer should get joined and simulated at the end, after the `!did_exit`, (which should only be set if `Esc` is pressed)

Also, people were reporting this happening in the middle of text, not at the end... Can you spoon feed me why it fixes it? 😅

File: src-tauri/src/main.rs (hunk 1 above)
Comment (kengoodridge) [reply src-tauri/src/main.rs:333]: gotta say all the things I'd actually want to use this for are text areas and not fields and are terribly crippled by no new line.

**Comment by kengoodridge** [comment]
> > no killing here. Wasn't super confident with it. Tell me if you can reproduce it tho.
> > Feel free to edit that out of this PR. The top part fixes the missing text tho, or seem to anyway.
> 
> I'm trying to understand why your changeset fixes it. The buffer should get joined and simulated at the end, after the `!did_exit`, (which should only be set if `Esc` is pressed)
> 
> Also, people were reporting this happening in the middle of text, not at the end... Can you spoon feed me why it fixes it? 😅

It's streaming in blocks of text, so any time less than 4 characters streams in it just skips it. It can be anywhere in the text. Not at the end.

File: src-tauri/src/main.rs (hunk 1 above)
Comment (jasonjmcghee) [reply src-tauri/src/main.rs:333]: Easy to fix by adding a user setting (allow enabling "simulate newlines with Enter key").

Strange that you don't have newlines though - they seem to work for me.

What system are you on?

Comment (jasonjmcghee) [reply src-tauri/src/main.rs:333]: (for example https://www.loom.com/share/d3adddc37c2e4c8daeef11255c0135ea?sid=5ec9ee09-5b9d-4222-aeff-f82cf241db74)

**Comment by jasonjmcghee** [comment]
> It's streaming in blocks of text, so any time less than 4 characters streams in it just skips it. It can be anywhere in the text. Not at the end necessarily.

But it adds it to the buffer, which isn't cleared, so it should still be there for the next iteration once it has enough text 🤔 what am I missing 😂

File: src-tauri/src/main.rs [hunk 2]
```diff
@@ -329,19 +329,18 @@ fn trigger_action(
...
 
                         for step in trigger.next_steps.clone() {
                             if let Step::StreamTextToScreen = step {
-                                enigo.text(&delta_output).expect("Failed to type out text");
+                                // Split the output on new lines and send each part followed by an 'Enter' keypress.
+                                for step in trigger.next_steps.clone() {
+                                    if let Step::StreamTextToScreen = step {
+                                        enigo.text(&delta_output).expect("Failed to type out text");
+                                    }
+                                }
```
Comment (jasonjmcghee) [inline src-tauri/src/main.rs:343]: looks like this may have been duplicated on accident

File: src-tauri/src/main.rs (hunk 1 above)
Comment (kengoodridge) [reply src-tauri/src/main.rs:333]: mac

**Comment by jasonjmcghee** [comment]
I reproduced it with your changeset:

https://github.com/jasonjmcghee/plock/assets/1522149/19746ac1-1dba-40d0-befe-d5a3d4b9b72a

**Comment by jasonjmcghee** [comment]
Was it the newlines that fixed it? I will try that too

**Comment by jasonjmcghee** [comment]
For what it's worth, my current supposition on this issue is something like the system is i/o blocking and enigo can't detect that and fails to enter the text. i plan to log an issue in the `enigo` repo here in a sec

File: src-tauri/src/main.rs (hunk 2 above)
Comment (kengoodridge) [reply src-tauri/src/main.rs]: fixed that.

**Comment by kengoodridge** [comment]
Something is up with my "fix". It definitely introduced the new line problem. I'm going to sleep on it.

It's weird cuz, I tested it several times and diffed the text inserted against the text in the console. 

Hopefully, I've not wasted your time.

**Comment by jasonjmcghee** [comment]
> Something is up with my "fix". It definitely introduced the new line problem. I'm going to sleep on it.
> 
> It's weird cuz, I tested it several times and diffed the text inserted against the text in the console.
> 
> Hopefully, I've not wasted your time.

Of course you haven’t. You taking the time to work on a fix is absolutely awesome- regardless of whether it ends up working!

**Comment by kengoodridge** [comment]
@jasonjmcghee TLDR try out the latest version and look at the code.

**Pull request merged**